bftclient/include/bftclient/bft_client.h

```diff
@@ -0,0 +1,101 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License"). You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <memory>
+#include <optional>
+
+#include "communication/ICommunication.hpp"
+#include "Logger.hpp"
+#include "DynamicUpperLimitWithSimpleFilter.hpp"
+
+#include "bftclient/config.h"
+#include "matcher.h"
+#include "msg_receiver.h"
+#include "exception.h"
+
+using namespace std::chrono_literals;
+
+namespace bft::client {
+
+class Client {
+ public:
+  Client(std::unique_ptr<bft::communication::ICommunication> comm, const ClientConfig& config)
+      : communication_(std::move(comm)),
+        config_(config),
+        quorum_converter_(config_.all_replicas, config_.f_val, config_.c_val),
+        expected_commit_time_ms_(config_.retry_timeout_config.initial_retry_timeout.count(),
+                                 config_.retry_timeout_config.number_of_standard_deviations_to_tolerate,
+                                 config_.retry_timeout_config.max_retry_timeout.count(),
+                                 config_.retry_timeout_config.min_retry_timeout.count(),
+                                 config_.retry_timeout_config.samples_per_evaluation,
+                                 config_.retry_timeout_config.samples_until_reset,
+                                 config_.retry_timeout_config.max_increasing_factor,
+                                 config_.retry_timeout_config.max_decreasing_factor) {
+    communication_->setReceiver(config_.id.val, &receiver_);
+    communication_->Start();
+  }
+
+  void Stop() { communication_->Stop(); }
+
+  // Send a message where the reply gets allocated by the callee and returned in a vector.
+  // The message to be sent is moved into the caller to prevent unnecessary copies.
+  //
+  // Throws a BftClientException on error.
+  Reply Send(const WriteConfig& config, Msg&& request);
+  Reply Send(const ReadConfig& config, Msg&& request);
+
+  // Useful for testing. Shouldn't be relied on in production.
+  std::optional<ReplicaId> primary() { return primary_; }
+
+ private:
+  // Generic function for sending a read or write message.
+  Reply Send(const MatchConfig& match_config, const RequestConfig& request_config, Msg&& request, bool read_only);
+
+  // Wait for messages until we get a quorum or a retry timeout.
+  //
+  // Return a Reply on quorum, or std::nullopt on timeout.
+  std::optional<Reply> Wait();
```

Comment:
I think we are missing a proper error reply handling in our functionality. For example, in the case of conflicting writes, we don't need to retry an operation.

Reply:
Isn't this mechanism just  an error response that matches from a quorum of replicas? The client itself is unaware of the contents of messages and our current client replies don't include any sort of success or failure marker, so I don't know how we could do this otherwise. This implementation mirrors the SimpleClient behavior.

Reply:
Error responses could be non-retriable like conflicts detected during a post-execution stage. In our case, we retry all the requests no matter of what result they return. Just something to think about as a functionality improvement.

Reply:
Sure. My point was that there's no way to know from the ClientReplyMsgHeader if a request actually had an error. It's in the serialized response. So an error would have to be returned from a quorum of nodes, which would match, and stop the retry. The caller to Client::Send would handle it, not the client itself.

Reply:
My point is that we may want to add a new field to the ClientReplyMsgHeader specifying what the error was, at least if it is retriable or not.

Reply:
Ah okay! Definitely something to discuss later :)

Reply:
We'd still need to wait for at least F+1 of those same error responses though.

Reply:
Sure